internal/sourceanalysis/sourceanalysis.go

```diff
@@ -0,0 +1,83 @@
+package sourceanalysis
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/google/osv-scanner/internal/govulncheckshim"
+	"github.com/google/osv-scanner/internal/output"
+	"github.com/google/osv-scanner/pkg/models"
+)
+
+type MatchedVulnerability struct {
+	vuln  models.Vulnerability
+	group *models.GroupInfo
+	pkg   models.PackageInfo
+}
+
+// vulnsFromAllPkgs returns a map of IDs to GroupInfo pointers, and the extracted list of vulnerabilities
+func vulnsFromAllPkgs(pkgs []models.PackageVulns) (map[string]map[string]MatchedVulnerability, []models.Vulnerability) {
+	idMatchedVulnMap := map[string]map[string]MatchedVulnerability{}
+	flatVulns := map[string]models.Vulnerability{}
+	for _, pv := range pkgs {
+		for _, vuln := range pv.Vulnerabilities {
+			groupIdx := slices.IndexFunc(pv.Groups, func(g models.GroupInfo) bool {
+				return slices.Contains(g.IDs, vuln.ID)
+			})
+
+			if idMatchedVulnMap[pv.Package.Name] == nil {
+				idMatchedVulnMap[pv.Package.Name] = make(map[string]MatchedVulnerability)
+			}
+
+			idMatchedVulnMap[pv.Package.Name][vuln.ID] = MatchedVulnerability{
+				group: &pv.Groups[groupIdx],
+				vuln:  vuln,
+				pkg:   pv.Package,
+			}
+
+			flatVulns[vuln.ID] = vuln
+		}
+	}
+
+	vulnList := []models.Vulnerability{}
+	for _, v := range flatVulns {
+		vulnList = append(vulnList, v)
+	}
+
+	return idMatchedVulnMap, vulnList
+}
+
+// Run runs the language specific analyzers on the code given packages and source info
+func Run(r *output.Reporter, source models.SourceInfo, pkgs []models.PackageVulns) {
+	idMatchedVulnMap, vulnsSlice := vulnsFromAllPkgs(pkgs)
+
+	// GoVulnCheck
+	if source.Type == "lockfile" && filepath.Base(source.Path) == "go.mod" {
+		res, err := govulncheckshim.RunGoVulnCheck(filepath.Dir(source.Path), vulnsSlice)
+		if err != nil {
+			// TODO: Better method to identify the type of error and give advice specific to the error
+			r.PrintError(
+				fmt.Sprintf("Failed to run code analysis (govulncheck) on '%s' because %s\n"+
+					"(the Go toolchain is required)\n", source.Path, err.Error()))
+
+			return
+		}
+		// Add analysis information back into package list
+		for _, v := range res.Vulns {
+			for _, m := range v.Modules {
+				if idMatchedVulnMap[m.Path] == nil {
```

Comment:
Looking into this a bit more. The `Vuln` struct is defined as:

```golang
type Vuln struct {
	// OSV contains all data from the OSV entry for this vulnerability.
	OSV *osv.Entry

	// Modules contains all of the modules in the OSV entry where a
	// vulnerable package is imported by the target source code or binary.
	//
	// For example, a module M with two packages M/p1 and M/p2, where only p1
	// is vulnerable, will appear in this list if and only if p1 is imported by
	// the target source code or binary.
	Modules []*Module
}
```

Assuming I'm understanding this correctly, let's take a project with lockfile that contains modules M1, M2, both of which are affected by OSV-1. Inside the project, let's say that we call vulnerable packages in M1, but only packages without vulnerabilities in M2. 

This means that govulncheck will give us:

```json
{
   "OSV": { OSV-1 },
   "Modules": [ M1 ],
}
```

Since M2 doesn't have any vulnerable packages imported. 

With the current code, this means that with our scan results, we find: OSV-1 affecting both M1 and M2, but we **don't** write `called: false` for M2 because that module is not in the `Modules` slice at all. This seems incorrect. 

I think instead we should be iterating over the modules, vulns that we have from our initial scan and cross referencing that with the results that govulncheck gives us to check the `IsCalled()`. We should also set `called` to `false` if a vulnerability was considered for source analysis but govulncheck did not return it or include the module (i.e. if it's not imported).

One approach that might be cleaner than all of this is something like:

```go
// Get rid of vulnsFromAllPkgs, MatchedVulnerability, and just iterate every single group we see in PackageVulns

// pre-process the govulncheck `res` to enable faster lookup. 

// slice naming confusing since a "package" in OSV-Schema actually refers to a Go module. 
for _, pkg :=  range pkgs {
  for _, groups := range pkg.Groups {
    for _, id := groups {
      // if the vulnerability for ID does not have source level info, continue
      // if the vulnerability for ID has source level info, check if it's in the listed vulns/modules returned by govulncheck. If not, set `called` to False. 
      // Set `called` to v.IsCalled()
     }
  }
}
```

Reply:
Rewrote source analysis according to the suggested approach, PTAL.